File: Signal/src/ViewControllers/DebugUI/DebugUIMessagesAction.m
//
//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
//

#import "DebugUIMessagesAction.h"
#import <SignalServiceKit/OWSPrimaryStorage.h>

NS_ASSUME_NONNULL_BEGIN

@class DebugUIMessagesSingleAction;

@interface DebugUIMessagesAction ()

@end

#pragma mark -

@interface DebugUIMessagesSingleAction ()

@property (nonatomic, nullable) ActionPrepareBlock prepareBlock;

// "Single" actions should have exactly one "staggered" or "unstaggered" action block.
@property (nonatomic, nullable) StaggeredActionBlock staggeredActionBlock;
@property (nonatomic, nullable) UnstaggeredActionBlock unstaggeredActionBlock;

@end

#pragma mark -

@implementation DebugUIMessagesAction

- (DebugUIMessagesSingleAction *)nextActionToPerform
{
    return (DebugUIMessagesSingleAction *)self;
}

- (void)prepare:(ActionSuccessBlock)success failure:(ActionFailureBlock)failure
{
    OWSAssertDebug(success);
    OWSAssertDebug(failure);

    OWSAbstractMethod();

    success();
}

- (void)prepareAndPerformNTimes:(NSUInteger)count
{
    OWSLogInfo(@"%@ prepareAndPerformNTimes: %zd", self.label, count);
    [DDLog flushLog];

    [self prepare:^{
        [self performNTimes:count
                    success:^{
                    }
                    failure:^{
                    }];
    }
          failure:^{
          }];
}

- (void)performNTimes:(NSUInteger)countParam success:(ActionSuccessBlock)success failure:(ActionFailureBlock)failure
{
    OWSAssertDebug(success);
    OWSAssertDebug(failure);

    OWSLogInfo(@"%@ performNTimes: %zd", self.label, countParam);
    [DDLog flushLog];

    if (countParam < 1) {
        success();
        return;
    }

    __block NSUInteger count = countParam;
    [OWSPrimaryStorage.sharedManager.newDatabaseConnection readWriteWithBlock:^(
        YapDatabaseReadWriteTransaction *transaction) {
        NSUInteger batchSize = 0;
        while (count > 0) {
            NSUInteger index = count;

            DebugUIMessagesSingleAction *action = [self nextActionToPerform];
            OWSAssertDebug([action isKindOfClass:[DebugUIMessagesSingleAction class]]);

            if (action.staggeredActionBlock) {
                OWSAssertDebug(!action.unstaggeredActionBlock);
                action.staggeredActionBlock(index,
                    transaction,
                    ^{
                        dispatch_after(
                            dispatch_time(DISPATCH_TIME_NOW, (int64_t)1.f * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                                OWSLogInfo(@"%@ performNTimes success: %zd", self.label, count);
                                [self performNTimes:count - 1 success:success failure:failure];
                            });
                    },
                    failure);

                break;
            } else {
                OWSAssertDebug(action.unstaggeredActionBlock);

                // TODO: We could check result for failure.
                action.unstaggeredActionBlock(index, transaction);

                const NSUInteger kMaxBatchSize = 2500;
                batchSize++;
                if (batchSize >= kMaxBatchSize) {
                    dispatch_after(
                        dispatch_time(DISPATCH_TIME_NOW, (int64_t)1.f * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                            OWSLogInfo(@"%@ performNTimes success: %zd", self.label, count);
                            [self performNTimes:count - 1 success:success failure:failure];
                        });

                    break;
                }
                count--;
            }
        }
    }];
}

@end

#pragma mark -

@implementation DebugUIMessagesSingleAction

+ (DebugUIMessagesAction *)actionWithLabel:(NSString *)label
                      staggeredActionBlock:(StaggeredActionBlock)staggeredActionBlock
{
    OWSAssertDebug(label.length > 0);
    OWSAssertDebug(staggeredActionBlock);

    DebugUIMessagesSingleAction *instance = [DebugUIMessagesSingleAction new];
    instance.label = label;
    instance.staggeredActionBlock = staggeredActionBlock;
    return instance;
}

+ (DebugUIMessagesAction *)actionWithLabel:(NSString *)label
                    unstaggeredActionBlock:(UnstaggeredActionBlock)unstaggeredActionBlock
{
    OWSAssertDebug(label.length > 0);
    OWSAssertDebug(unstaggeredActionBlock);

    DebugUIMessagesSingleAction *instance = [DebugUIMessagesSingleAction new];
    instance.label = label;
    instance.unstaggeredActionBlock = unstaggeredActionBlock;
    return instance;
}

+ (DebugUIMessagesAction *)actionWithLabel:(NSString *)label
                      staggeredActionBlock:(StaggeredActionBlock)staggeredActionBlock
                              prepareBlock:(ActionPrepareBlock)prepareBlock
{
    OWSAssertDebug(label.length > 0);
    OWSAssertDebug(staggeredActionBlock);
    OWSAssertDebug(prepareBlock);

    DebugUIMessagesSingleAction *instance = [DebugUIMessagesSingleAction new];
    instance.label = label;
    instance.staggeredActionBlock = staggeredActionBlock;
    instance.prepareBlock = prepareBlock;
    return instance;
}

+ (DebugUIMessagesAction *)actionWithLabel:(NSString *)label
                    unstaggeredActionBlock:(UnstaggeredActionBlock)unstaggeredActionBlock
                              prepareBlock:(ActionPrepareBlock)prepareBlock
{
    OWSAssertDebug(label.length > 0);
    OWSAssertDebug(unstaggeredActionBlock);
    OWSAssertDebug(prepareBlock);

    DebugUIMessagesSingleAction *instance = [DebugUIMessagesSingleAction new];
    instance.label = label;
    instance.unstaggeredActionBlock = unstaggeredActionBlock;
    instance.prepareBlock = prepareBlock;
    return instance;
}

- (void)prepare:(ActionSuccessBlock)success failure:(ActionFailureBlock)failure
{
    OWSAssertDebug(success);
    OWSAssertDebug(failure);

    if (self.prepareBlock) {
        self.prepareBlock(success, failure);
    } else {
        success();
    }
}

@end

#pragma mark -

@interface DebugUIMessagesGroupAction ()

@property (nonatomic) SubactionMode subactionMode;
@property (nonatomic, nullable) NSArray<DebugUIMessagesAction *> *subactions;
@property (nonatomic) NSUInteger subactionIndex;

@end

#pragma mark -

@implementation DebugUIMessagesGroupAction

- (DebugUIMessagesSingleAction *)nextActionToPerform
{
    OWSAssertDebug(self.subactions.count > 0);

    switch (self.subactionMode) {
        case SubactionMode_Random: {
            DebugUIMessagesAction *subaction = self.subactions[arc4random_uniform((uint32_t)self.subactions.count)];
            OWSAssertDebug(subaction);
            return subaction.nextActionToPerform;
        }
        case SubactionMode_Ordered: {
            DebugUIMessagesAction *subaction = self.subactions[self.subactionIndex];
            OWSAssertDebug(subaction);
            self.subactionIndex = (self.subactionIndex + 1) % self.subactions.count;
            return subaction.nextActionToPerform;
        }
    }
}

- (void)prepare:(ActionSuccessBlock)success failure:(ActionFailureBlock)failure
{
    OWSAssertDebug(success);
    OWSAssertDebug(failure);

    [DebugUIMessagesGroupAction prepareSubactions:[self.subactions mutableCopy] success:success failure:failure];
}

+ (void)prepareSubactions:(NSMutableArray<DebugUIMessagesAction *> *)unpreparedSubactions
                  success:(ActionSuccessBlock)success
                  failure:(ActionFailureBlock)failure
{
    OWSAssertDebug(success);
    OWSAssertDebug(failure);

    if (unpreparedSubactions.count < 1) {
        return success();
    }

    DebugUIMessagesAction *nextAction = unpreparedSubactions.lastObject;
    [unpreparedSubactions removeLastObject];
    OWSLogInfo(@"preparing: %@", nextAction.label);
    [DDLog flushLog];
    [nextAction prepare:^{
        [self prepareSubactions:unpreparedSubactions success:success failure:failure];
    }
                failure:^{
                }];
}

+ (DebugUIMessagesAction *)randomGroupActionWithLabel:(NSString *)label
                                           subactions:(NSArray<DebugUIMessagesAction *> *)subactions
{
    OWSAssertDebug(label.length > 0);
    OWSAssertDebug(subactions.count > 0);

    DebugUIMessagesGroupAction *instance = [DebugUIMessagesGroupAction new];
    instance.label = label;
    instance.subactions = subactions;
    instance.subactionMode = SubactionMode_Random;
    return instance;
}

+ (DebugUIMessagesAction *)allGroupActionWithLabel:(NSString *)label
                                        subactions:(NSArray<DebugUIMessagesAction *> *)subactions
{
    OWSAssertDebug(label.length > 0);
    OWSAssertDebug(subactions.count > 0);

    DebugUIMessagesGroupAction *instance = [DebugUIMessagesGroupAction new];
    instance.label = label;
    instance.subactions = subactions;
    instance.subactionMode = SubactionMode_Ordered;
    return instance;
}

@end

NS_ASSUME_NONNULL_END

File: Signal/src/ViewControllers/DebugUI/DebugUIMessagesAction.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalServiceKit

#if USE_DEBUG_UI

class DebugUIMessagesAction: Dependencies {

    typealias Completion = (Result<Void, Error>) -> Void

    let label: String

    fileprivate init(label: String) {
        self.label = label
    }

    fileprivate func nextActionToPerform() -> DebugUIMessagesSingleAction {
        return self as! DebugUIMessagesSingleAction
    }

    fileprivate func prepare(completion: @escaping Completion) {
        completion(.success(()))
    }

    func prepareAndPerformNTimes(_ count: UInt) {
        prepare { result in
            switch result {
            case .success:
                self.performNTimes(count, completion: { _ in })
            case .failure:
                break
            }
        }
    }

    private func performNTimes(_ count: UInt, completion: @escaping Completion) {
        Logger.info("\(label) performNTimes: \(count)")
        Logger.flush()

        guard count > 0 else {
            completion(.success(()))
            return
        }

        var runCount = count
        databaseStorage.write { transaction in
            var batchSize = 0
            while runCount > 0 {
                let index = runCount

                let action = nextActionToPerform()
                if let staggeredAction = action.staggeredAction {
                    owsAssertDebug(action.unstaggeredAction == nil)
                    staggeredAction(index, transaction, { result in
                        switch result {
                        case .success:
                            DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                                Logger.info("\(self.label) performNTimes success: \(runCount)")
                                self.performNTimes(runCount - 1, completion: completion)
                            }
                        case .failure(let error):
                            completion(.failure(error))
                        }
                    })
                    break
                } else if let unstaggeredAction = action.unstaggeredAction {
                    // TODO: We could check result for failure.
                    unstaggeredAction(index, transaction)

                    let maxBatchSize = 2500
                    batchSize += 1
                    if batchSize >= maxBatchSize {
                        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                            Logger.info("\(self.label) performNTimes success: \(runCount)")
                            self.performNTimes(runCount - 1, completion: completion)
                        }
                        break
                    }
                    runCount -= 1
                } else {
                    owsFailDebug("No staggeredActionBlock / unstaggeredActionBlock")
                }
            }
        }
    }
}

class DebugUIMessagesSingleAction: DebugUIMessagesAction {

    typealias StaggeredAction = (UInt, SDSAnyWriteTransaction, @escaping Completion) -> Void
    typealias UnstaggeredAction = (UInt, SDSAnyWriteTransaction) -> Void

    private(set) var prepare: ((@escaping Completion) -> Void)?
    private(set) var staggeredAction: StaggeredAction?
    private(set) var unstaggeredAction: UnstaggeredAction?

    init(label: String, staggeredAction: @escaping StaggeredAction, prepare: ((@escaping Completion) -> Void)? = nil) {
        super.init(label: label)
        self.staggeredAction = staggeredAction
        self.prepare = prepare
    }

    init(label: String, unstaggeredAction: @escaping UnstaggeredAction, prepare: ((@escaping Completion) -> Void)? = nil) {
        super.init(label: label)
        self.unstaggeredAction = unstaggeredAction
        self.prepare = prepare
    }

    override func prepare(completion: @escaping Completion) {
        guard let prepare else {
            completion(.success(()))
            return
        }
        prepare(completion)
    }
}

class DebugUIMessagesGroupAction: DebugUIMessagesAction {

    enum SubactionMode {
        case random
        case ordered
    }
    let mode: SubactionMode
    let subactions: [DebugUIMessagesAction]
    private var subactionIndex: Array.Index = 0

    private init(label: String, subactions: [DebugUIMessagesAction], mode: SubactionMode) {
        self.subactions = subactions
        self.mode = mode
        super.init(label: label)
        subactionIndex = subactions.startIndex
    }

    // Given a group of subactions, perform a single random subaction each time.
    final class func randomGroupActionWithLabel(_ label: String, subactions: [DebugUIMessagesAction]) -> DebugUIMessagesAction {
        return DebugUIMessagesGroupAction(label: label, subactions: subactions, mode: .random)
    }

    // Given a group of subactions, perform the subactions in order.
    //
    // If prepareAndPerformNTimes: is called with count == subactions.count, all of the subactions
    // are performed exactly once.
    final class func allGroupActionWithLabel(_ label: String, subactions: [DebugUIMessagesAction]) -> DebugUIMessagesAction {
        return DebugUIMessagesGroupAction(label: label, subactions: subactions, mode: .ordered)
    }

    override func nextActionToPerform() -> DebugUIMessagesSingleAction {
        let subaction: DebugUIMessagesAction = {
            switch mode {
            case .random:
                return subactions.randomElement()!

            case .ordered:
                let subaction = subactions[subactionIndex]
                if subactionIndex < subactions.endIndex {
                    subactionIndex = subactions.index(after: subactionIndex)
                } else {
                    subactionIndex = subactions.startIndex
                }
                return subaction
            }
        }()
        return subaction.nextActionToPerform()
    }

    override func prepare(completion: @escaping Completion) {
        DebugUIMessagesGroupAction.prepareSubactions(subactions, completion: completion)
    }

    private static func prepareSubactions(_ subactions: [DebugUIMessagesAction], completion: @escaping Completion) {
        guard !subactions.isEmpty else {
            completion(.success(()))
            return
        }

        var unpreparedSubactions = subactions
        let nextAction = unpreparedSubactions.popLast()!
        Logger.info("Preparing: \(nextAction.label)")
        Logger.flush()
        nextAction.prepare { result in
            switch result {
            case .success:
                self.prepareSubactions(unpreparedSubactions, completion: completion)

            case .failure:
                break
            }
        }
    }

}

#endif
